fix: do not remove webhooks during initialization

johnbuluba opened the pull request:
## Related issue

Closes #3625

@realshuting 

## Milestone of this PR

/milestone 1.7.0

## What type of PR is this

/kind bug

## Proposed Changes

Currently during the webhook registration leader initialization all webhook configs are deleted and recreated. This creates a small time window that the cluster doesn't have any webhook configuration and as a result the resources are not validated.

This PR changes the registration logic to not remove the webhook configurations on the leader initialization. Instead it updates the existing webhook configuration to the current desired values. 

### Proof Manifests

To test this PR:
1. Create a Policy with `validationFailureAction: enforce` :

   ```yaml
   apiVersion: kyverno.io/v1
   kind: Policy
   metadata:
     name: disallow-host-namespaces
   spec:
     validationFailureAction: enforce
     background: true
     rules:
       - name: host-namespaces
         match:
           any:
           - resources:
               kinds:
                 - Pod
         validate:
           message: >-
             Sharing the host namespaces is disallowed. The fields spec.hostNetwork,
             spec.hostIPC, and spec.hostPID must be unset or set to `false`.          
           pattern:
             spec:
               =(hostPID): "false"
               =(hostIPC): "false"
               =(hostNetwork): "false"
   ```

2. Create an invalid resource based on the previous Policy, e.g. create a `pod.yaml`: 

   ```yaml
   apiVersion: v1
   kind: Pod
   metadata:
     name: privilegedr
   spec:
     hostPID: true
     hostNetwork: true
     containers:
     - name: alpine
       image: alpine:3.7
       command:
       - nsenter
       - "--mount=/proc/1/ns/mnt"
       - "--"
       - sh
       - "-"
       stdin: true
       tty: true
       resources:
         requests:
           cpu: 10m
       securityContext:
         privileged: true
   ```

3. Start applying the in a loop: 
   ```sh
   until kubectl apply -f pod.yaml; do echo "trying again"; done
   Error from server: error when creating "pod.yaml": admission webhook "validate.kyverno.svc-fail" denied the request: 
   
   resource Pod/kubeflow-user/privileged was blocked due to the following policies
   
   disallow-host-namespaces: 
 	  host-namespaces: 'validation error: Sharing the host namespaces is disallowed. The
   fields spec.hostNetwork, spec.hostIPC, and spec.hostPID must be unset or set to
   `false`.          . Rule host-namespaces failed at path /spec/hostNetwork/'
   trying again
   ```

4. In a new terminal delete the Kyverno pod: 
     ```bash
   kubectl delete pods -n kyverno --all
   ```

5. Wait until the new Kyverno pod initializes. When it does that, the command in step 3 will succeed and create the resource. 

This PR should fix this behavior and the command in step 3 will not succeed.

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->

- [x] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [x] I have read the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) and followed the process including adding proof manifests to this PR.
- [ ] This is a bug fix and I have added unit tests that prove my fix is effective.
- [ ] This is a feature and I have added CLI tests that are applicable.
- [ ] My PR needs to be cherry picked to a specific release branch which is <replace>.
- [ ] My PR contains new or altered behavior to Kyverno and
  - [ ] CLI support should be added and my PR doesn't contain that functionality.
  - [ ] I have added or changed [the documentation](https://github.com/kyverno/website) myself in an existing PR and the link is:
  <!-- Uncomment to link to the PR -->
  <!-- https://github.com/kyverno/website/pull/123 -->
  - [ ] I have raised an issue in [kyverno/website](https://github.com/kyverno/website) to track the documentation update and the link is:
  <!-- Uncomment to link to the issue -->
  <!-- https://github.com/kyverno/website/issues/1 -->

## Further Comments

The solution is not as simple as just removing the `wrc.removeWebhookConfigurations()` from the `Register` method. The reson for this is mainly the `autoUpdateWebhooks` flag.

More specifically if on the first run Kyverno has the flag `autoUpdateWebhooks: false` it will create a "catch all" `ValidatingWebhookConfiguration` for the resources with one webhook defined. 

If we later change the `autoUpdateWebhooks: true`, the webhooks are not reinitialized, thus the `ValidatingWebhookConfiguration` becomes out of sync with only the `validate.kyverno.svc-ignore` webhook defined. When the Config Manager is trying to update the webhook configuration, it will not find `validate.kyverno.svc-fail` and not apply the correct rules.

This PR solves it by implementing 2 methods in `registration`: 
1. `updateValidatingWebhookConfiguration`
1. `updateMutatingWebhookConfiguration`

These methods are updating the existing webhook configurations during initialization.  If the `autoUpdateWebhooks: true`, the existing rules will not be ovewritten and  the `validate.kyverno.svc-fail` resources will fail until the Kyverno Pod is up.
Also we need to syncrhonize the `caBundle` value to the existing webhook configuration, in case it changed.


An extra note is that there was a minor issue in the `webhookRulesEqual` function, that when the API had 1 value but the internal representation was empty it returned `true`. The reason for that is that the `len` of the 2 arrays was 1, but the internal representation had one empty object.

vyankyGH commented:
Hey @johnbuluba Thank for submitting PR.
There are few comments on PR can you please address?

- Instead of using dynamic client we can use typed client. If we used dynamic client and updated so frequently will get throttle for those request either GET or POST.
- as Dynamic client shared entire kyverno package, so we should avoid  using if it possible.
- For the get resource can use Lister and to update resource can the typed client.

Also the E2e test also failing can you please look at it?

Thanks.
cc @realshuting

johnbuluba commented:
@vyankyGH 

Thanks for your review!

I can update the logic to use Lister, but I am having trouble finding usage in the codebase that use typed client. Should I extend the client? 
Is there an example in the codebase that use the typed client that I can use for guidance? 

I will take a look at the tests errors.


Thanks

vyankyGH commented:
> Thanks for your review @vyankyGH !
> 
> I can update the logic to use Lister, but I am having trouble finding usage in the codebase that use typed client. Should I extend the client? Is there an example in the codebase that use the typed client that I can use for guidance?
> 
> I will take a look at the tests errors.
> 
> Thanks

Hey @johnbuluba we have check there’s no webhook client in webhook register. client.kclient is private which cannot be exposed.
so for now it's okay to used dynamic client.

johnbuluba commented:
@vyankyGH I updated the PR to: 

1. Use the Lister instead of the dynamic client to fetch the `ValidatingWebhookConfiguration` and `MutatingWebhookConfiguration`.
2. I removed the commit that fixed the issue with the `webhookRulesEqual` function as it is unrelated to the original issue and it was causing the E2E tests error.

I run the tests locally by running: 
```bash
make kind-e2e-cluster
make kind-deploy
make test
```

Please take a look for any other improvements.

Thanks!

johnbuluba commented:
A sidenote regarding this commit 6a3fe6f89ed4c201c7c726e67852a2a2f1e65cd7 that I removed about this issue: 

> An extra note is that there was a minor issue in the webhookRulesEqual function, that when the API had 1 value but the internal representation was empty it returned true. The reason for that is that the len of the 2 arrays was 1, but the internal representation had one empty object.

I think this is a valid bug, you can reproduce it with the following steps: 
1. Apply a mutating policy, for example : 
   ```yaml
   apiVersion: kyverno.io/v1
   kind: ClusterPolicy
   metadata:
     name: add-volume
     annotations:
       policies.kyverno.io/title: Add Volume to Deployment
       policies.kyverno.io/category: Sample
       policies.kyverno.io/subject: Deployment, Volume
       policies.kyverno.io/description: >-
         Some Kubernetes applications like HashiCorp Vault must perform some modifications
         to resources in order to invoke their specific functionality. Often times, that functionality
         is controlled by the presence of a label or specific annotation. This policy, based on HashiCorp
         Vault, adds a volume and volumeMount to a Deployment if there is an annotation called
         "vault.k8s.corp.net/inject=enabled" present.
   spec:
     rules:
     - name: add-volume
       match:
         resources:
           kinds:
           - Deployment
       preconditions:
         any:
         - key: "{{request.object.spec.template.metadata.annotations.\"vault.k8s.corp.net/inject\"}}"
           operator: Equals
           value: "enabled"
       mutate:
         patchesJson6902: |-
           - op: add
             path: /spec/template/spec/volumes
             value: [{"name": "vault-secret","emptyDir": {"medium": "Memory"}}]
   ```
2. Verify that the `kyverno-resource-mutating-webhook-cfg` configuration rules are updated:
   ```yaml
   ...
    rules:
    - apiGroups:
      - apps
      apiVersions:
      - v1
      operations:
      - CREATE
      - UPDATE
      - DELETE
      resources:
      - deployments
      scope: '*'
   ...
   ``` 
3. Delete the policy you created in step 1: 
   ```bash
   kubectl delete clusterpolicy add-volume
   ```
4. Verify that the `kyverno-resource-mutating-webhook-cfg` configuration rules are **not** updated:
   ```yaml
   ...
    rules:
    - apiGroups:
      - apps
      apiVersions:
      - v1
      operations:
      - CREATE
      - UPDATE
      - DELETE
      resources:
      - deployments
      scope: '*'
   ...
   ``` 

This happens only in the `main` branch. In version `1.6.2` this worked as expected, i.e. in step 4 the rules were empty. 

If you think that this is a bug, I can open a new issue and tackle this separately .

vyankyGH commented:
> @vyankyGH I updated the PR to:
> 
> 1. Use the Lister instead of the dynamic client to fetch the `ValidatingWebhookConfiguration` and `MutatingWebhookConfiguration`.
> 2. I removed the commit that fixed the issue with the `webhookRulesEqual` function as it is unrelated to the original issue and it was causing the E2E tests error.
> 
> I run the tests locally and everything seems OK now. I run them with:
> 
> ```shell
> make kind-e2e-cluster
> make kind-deploy
> make test
> ```
> 
> Please take a look for any other improvements.
> 
> Thanks!

Thanks @johnbuluba will take a look

vyankyGH commented:
> A sidenote regarding this commit [6a3fe6f](https://github.com/kyverno/kyverno/commit/6a3fe6f89ed4c201c7c726e67852a2a2f1e65cd7) that I removed about this issue:
> 
> > An extra note is that there was a minor issue in the webhookRulesEqual function, that when the API had 1 value but the internal representation was empty it returned true. The reason for that is that the len of the 2 arrays was 1, but the internal representation had one empty object.
> 
> I think this is a valid bug, you can reproduce it with the following steps:
> 
> 1. Apply a mutating policy, for example :
>    ```yaml
>    apiVersion: kyverno.io/v1
>    kind: ClusterPolicy
>    metadata:
>      name: add-volume
>      annotations:
>        policies.kyverno.io/title: Add Volume to Deployment
>        policies.kyverno.io/category: Sample
>        policies.kyverno.io/subject: Deployment, Volume
>        policies.kyverno.io/description: >-
>          Some Kubernetes applications like HashiCorp Vault must perform some modifications
>          to resources in order to invoke their specific functionality. Often times, that functionality
>          is controlled by the presence of a label or specific annotation. This policy, based on HashiCorp
>          Vault, adds a volume and volumeMount to a Deployment if there is an annotation called
>          "vault.k8s.corp.net/inject=enabled" present.
>    spec:
>      rules:
>      - name: add-volume
>        match:
>          resources:
>            kinds:
>            - Deployment
>        preconditions:
>          any:
>          - key: "{{request.object.spec.template.metadata.annotations.\"vault.k8s.corp.net/inject\"}}"
>            operator: Equals
>            value: "enabled"
>        mutate:
>          patchesJson6902: |-
>            - op: add
>              path: /spec/template/spec/volumes
>              value: [{"name": "vault-secret","emptyDir": {"medium": "Memory"}}]
>    ```
> 2. Verify that the `kyverno-resource-mutating-webhook-cfg` configuration rules are updated:
>    ```yaml
>    ...
>     rules:
>     - apiGroups:
>       - apps
>       apiVersions:
>       - v1
>       operations:
>       - CREATE
>       - UPDATE
>       - DELETE
>       resources:
>       - deployments
>       scope: '*'
>    ...
>    ```
> 3. Delete the policy you created in step 1:
>    ```shell
>    kubectl delete clusterpolicy add-volume
>    ```
> 4. Verify that the `kyverno-resource-mutating-webhook-cfg` configuration rules are **not** updated:
>    ```yaml
>    ...
>     rules:
>     - apiGroups:
>       - apps
>       apiVersions:
>       - v1
>       operations:
>       - CREATE
>       - UPDATE
>       - DELETE
>       resources:
>       - deployments
>       scope: '*'
>    ...
>    ```
> 
> This happens only in the `main` branch. In version `1.6.2` this worked as expected, i.e. in step 4 the rules were empty.
> 
> If you think that this is a bug, I can open a new issue and tackle this separately .

Yes please please log an issue.

vyankyGH commented:
/LGTM
@realshuting any more comment's?

realshuting reviewed (approved):
/lgtm 

Thanks for your contribution @johnbuluba !

realshuting commented:
> A sidenote regarding this commit [6a3fe6f](https://github.com/kyverno/kyverno/commit/6a3fe6f89ed4c201c7c726e67852a2a2f1e65cd7) that I removed about this issue:
> 
> > An extra note is that there was a minor issue in the webhookRulesEqual function, that when the API had 1 value but the internal representation was empty it returned true. The reason for that is that the len of the 2 arrays was 1, but the internal representation had one empty object.
> 
> I think this is a valid bug, you can reproduce it with the following steps:
> 
> 1. Apply a mutating policy, for example :
>    ```yaml
>    apiVersion: kyverno.io/v1
>    kind: ClusterPolicy
>    metadata:
>      name: add-volume
>      annotations:
>        policies.kyverno.io/title: Add Volume to Deployment
>        policies.kyverno.io/category: Sample
>        policies.kyverno.io/subject: Deployment, Volume
>        policies.kyverno.io/description: >-
>          Some Kubernetes applications like HashiCorp Vault must perform some modifications
>          to resources in order to invoke their specific functionality. Often times, that functionality
>          is controlled by the presence of a label or specific annotation. This policy, based on HashiCorp
>          Vault, adds a volume and volumeMount to a Deployment if there is an annotation called
>          "vault.k8s.corp.net/inject=enabled" present.
>    spec:
>      rules:
>      - name: add-volume
>        match:
>          resources:
>            kinds:
>            - Deployment
>        preconditions:
>          any:
>          - key: "{{request.object.spec.template.metadata.annotations.\"vault.k8s.corp.net/inject\"}}"
>            operator: Equals
>            value: "enabled"
>        mutate:
>          patchesJson6902: |-
>            - op: add
>              path: /spec/template/spec/volumes
>              value: [{"name": "vault-secret","emptyDir": {"medium": "Memory"}}]
>    ```
> 2. Verify that the `kyverno-resource-mutating-webhook-cfg` configuration rules are updated:
>    ```yaml
>    ...
>     rules:
>     - apiGroups:
>       - apps
>       apiVersions:
>       - v1
>       operations:
>       - CREATE
>       - UPDATE
>       - DELETE
>       resources:
>       - deployments
>       scope: '*'
>    ...
>    ```
> 3. Delete the policy you created in step 1:
>    ```shell
>    kubectl delete clusterpolicy add-volume
>    ```
> 4. Verify that the `kyverno-resource-mutating-webhook-cfg` configuration rules are **not** updated:
>    ```yaml
>    ...
>     rules:
>     - apiGroups:
>       - apps
>       apiVersions:
>       - v1
>       operations:
>       - CREATE
>       - UPDATE
>       - DELETE
>       resources:
>       - deployments
>       scope: '*'
>    ...
>    ```
> 
> This happens only in the `main` branch. In version `1.6.2` this worked as expected, i.e. in step 4 the rules were empty.
> 
> If you think that this is a bug, I can open a new issue and tackle this separately .

@johnbuluba - thanks for pointing this out! 

I've logged https://github.com/kyverno/kyverno/issues/3653 for tracking, can we get this fixed in 1.7?

vyankyGH commented:
> @vyankyGH I updated the PR to:
> 
> 1. Use the Lister instead of the dynamic client to fetch the `ValidatingWebhookConfiguration` and `MutatingWebhookConfiguration`.
> 2. I removed the commit that fixed the issue with the `webhookRulesEqual` function as it is unrelated to the original issue and it was causing the E2E tests error.
> 
> I run the tests locally and everything seems OK now. I run them with:
> 
> ```shell
> make kind-e2e-cluster
> make kind-deploy
> make test
> ```
> 
> Please take a look for any other improvements.
> 
> Thanks!

Thanks for your contribution @johnbuluba !
